### ISSUE TITLE
Docker daemon unresponsive when doing a big `docker cp`.

### ISSUE THREAD

**Issue by nschoe** [docker/docker#29058]
Hi,

I have made a database backup from one container running postgresql, and I used `docker cp` to copy the SQL dump out of the container to my host. The dump file was 7G and when the `docker cp` was in progress, in another terminal I ran `docker ps` and it hanged until `docker cp` terminated.

Is this normal?

For information:

```Bash
docker version
Client:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   b9f10c9
 Built:        Wed Jun  1 22:00:43 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   b9f10c9
 Built:        Wed Jun  1 22:00:43 2016
 OS/Arch:      linux/amd64
```

And: 
```Bash
docker info                                                                                                       nschoe@mimolette
Containers: 36
 Running: 27
 Paused: 0
 Stopped: 9
Images: 405
Server Version: 1.11.2
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 395
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: overlay host bridge null
Kernel Version: 4.4.0-31-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 62.81 GiB
Name: yyy
ID: xxx
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Cluster store: consul://127.0.0.1:8500
```

**Comment by thaJeztah** [comment]
ping @jlhawn @unclejack

**Comment by jlhawn** [comment]
I can repro this on `docker 1.13.0-rc2`.

It looks like both the `docker container cp ...` and `docker container ps` commands get a lock on the container object before doing the operations:

- [during `containerArchivePath()`](https://github.com/docker/docker/blob/fddb5a7f2ab29be441bc797f7e79fffc5cc531d3/daemon/archive.go#L107)
- [during `reducePsContainer()`](https://github.com/docker/docker/blob/fddb5a7f2ab29be441bc797f7e79fffc5cc531d3/daemon/list.go#L202)

I don't know *why* these technically have to be mutually exclusive but they are. I feel like a shared read lock would be more appropriate in this case but I'm not sure what the consequences would be of making such a change.

**Comment by cpuguy83** [comment]
Yes, this is a known issue since we are taking the container lock, which ensures the container isn't removed from underneath us while performing the copy.
I readlock would work, but this introduces a tremendous amount of complexity.

Currently working on just removing the container lock all-together.

**Comment by jlhawn** [comment]
> I have made a database backup from one container running postgresql, and I used docker cp to copy the SQL dump out of the container to my host.

As a temporary work-around, since it looks like you are running some secondary backup job in the container anyway, you could try a `docker exec ...` command which writes the backup file(s) to standard output. I don't believe this would hold a lock.

**Comment by thaJeztah** [comment]
> Currently working on just removing the container lock all-together.

For reference, https://github.com/docker/docker/pull/28483 is part of that effort
